File: Parser.java
import org.apache.pdfbox.pdmodel.*;
import org.apache.pdfbox.text.*;
import org.apache.pdfbox.pdmodel.interactive.documentnavigation.outline.PDOutlineItem;

import org.apache.pdfbox.cos.COSDictionary;

import java.io.*;

import java.util.*;

import java.util.regex.*;

public class Parser {

	public static void main(String[] args) {

		PDDocument doc = new PDDocument();
		String text="";

		try {

			File file = new File(args[0]+".pdf");   // input PDF file

			//File output = new File(args[1]+".txt");   // output text file

			doc = PDDocument.load(file);

			PDFTextStripper stripper = new PDFTextStripper();

			stripper.setSortByPosition(true);

			//stripper.setSpacingTolerance(0.9f);   // doesn't seem to make a difference?
			//stripper.setEndPage(1);

			text = stripper.getText(doc).replaceAll("(?m)^[ \t]*\r?\n", "");   // remove empty lines/whitespace

/*			BufferedWriter wr = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(output)));
			stripper.writeText(doc, wr);

			wr.close();*/

		} catch (Exception e) {
			System.out.println(e.getMessage());
		} finally {
   			if( doc != null) {
   				try {
    					doc.close();
    					System.out.println("Don't worry - done closing now!");
   				} catch (Exception e) {
   					System.out.println(e.getMessage());
   				}
   			}
   		}


   		// write text to file

   		BufferedWriter writer = null;
		try
		{
		    writer = new BufferedWriter(new FileWriter(args[0]+".txt"));
		    writer.write(text);

		}
		catch ( IOException e)
		{
		}
		finally
		{
		    try
		    {
		        if ( writer != null)
		        writer.close( );
		    }
		    catch ( IOException e)
		    {
		    }
		}

		// end

		/* doing this in python now
   		String[] lines = text.split("\n");
   		String line;

   		for (int i=0; i<lines.length;i++) {   // iterate over text line by line
   			line = lines[i].toLowerCase().trim();   // force case insensitivity, remove leading/trailing whitespace
   			//System.out.println(line);
   			if (line.matches(".*time.*:\\s*.*"))   // check if line might contain time details
   				parseTime(line);
   		*/


   		}


	}
	
	/* doing this in python now
	public static void parseTime(String next) {   // parses a line that is *suspected* to contain time details
		Pattern pattern = Pattern.compile(".*time.*:\\s*(?:([0-9]{1,2}(?:\\.[0-9]{0,2})?)\\s*(?:hours|hour|hrs|hr|hs|h))?\\s*([0-9]{1,3})?.*");   // regex to match time details
		Matcher matcher = pattern.matcher(next);
		matcher.matches();   // attempt to match

		String hrs = matcher.group(1);   // extract groups, capturing string representaions of hours and minutes
		String mins = matcher.group(2);

		if ((hrs == null) && (mins == null))   // nothing captured
			System.out.println("Sorry I thought I might be able to but I could not find a reasonable time duration from this line: " + next);
		else {
			hrs = (hrs == null) ? "0" : hrs;   // if no hours captured, set to 0
			mins = (mins == null) ? "0" : mins;   // if no mins captured, set to 0

			int h; int m;   // int representations of hrs and mins

			if (hrs.matches("[0-9]{1,2}\\.[0-9]{1,2}")) {   // if hours is currently a decimal
				h = Integer.parseInt((hrs.split("\\."))[0]);
				m = (int) ((Float.parseFloat("0."+(hrs.split("\\."))[1]))*60);   // split into hours and minutes
			} else {
				h = Integer.parseInt(hrs);
				m = Integer.parseInt(mins);
			}

			System.out.println("Time found: " + h + " hours and " + m + " minutes");
		}
	

	}

}

*/

File: parser.py
import sys, re

from pyparsing import Word, alphas

def main():

	# regex to match line that hopefully contains time details e.g. 'Time allowed: 1.5 hours.'
	time_regex = re.compile(r'.*time.*:\s*'   # match hint to time info i.e. '...time...:'
							r'([0-9.]{1,3}'   # open capture; match numbers including decimals
							r'\s{0,2}'   # match any spaces in between
							r'[a-z]+'   # match at least one of any word, hopefully things like 'hours','h','hrs'
							r'[^.]*)')   # match remainder of line excluding dots; close capture

	# regex to match line that hopefully contains calculator details e.g. 'Calculators are not allowed'
	calc_regex = re.compile(r'(?!.*\bno\b.*)'   # don't match if line contains word 'no' (otherwise might be fooled)
							r'.*calculators('   # match anything followed by word 'calculators'; open capture
							r'[^.]*)')   # match remainder of line excluding dots; close capture

	f = open(sys.argv[1]+".txt")   # open text file of past paper PDF

	for line in f:   # iterate over lines

		line = line.strip().lower()   # force case insensitivity

		time_match = time_regex.match(line)   # matches only at beginning of string
		if (time_match):
			parseTime(time_match.group(1))   # pass extracted details (e.g. '2 hours') to function to interpret

		calc_match = calc_regex.match(line)
		if (calc_match):
			parseCalc(calc_match.group(1))   # pass extracted details (e.g. 'are permitted') to function to interpret

	f.close()

	pass

def parseTime(line):   # function to interpret line containing time details, returning time allowed in terms of x hours and y minutes

	# match against a mighty regex to attempt to find the time allowed from the line, captured in groups hours and minutes
	
	hrWords = 'hours|hour|hrs|hr|hs|h'
	minWords = 'minutes|minute|mins|min|ms|m'

	hrmin_regex = re.compile(r'^(?:([0-9]{1,3}(?:\.[0-9]{0,2})?)'   # match numbers, optionally a decimal; capture in group 1 for hours
							 r'\s{0,2}'   # match optional spaces in between
							 r'(?:'+hrWords+'))?'   # match a word representing hours, whole hours part is optional
							 r'\s{0,2}'   # match optional spaces in between
							 r'(?:([0-9]{1,3})'   # match numbers, capture in group 2 for minutes
							 r'\s{0,2}'   # match optional spaces in between
							 r'(?:'+minWords+')?)?'   # match a word representing minutes, whole minutes part is optional
							 r'(?<=[0-9rsenmh])$')   # match only if string ends in one of the given characters (stops match empty string)

	hrmin_match = hrmin_regex.match(line)
	
	if (hrmin_match):
		hrs = hrmin_match.group(1)
		mins = hrmin_match.group(2)
		if (hrs != None):   # if hours are captured
			dec_match = re.match(r'([0-9]{1,2})\.([0-9]{1,2})',hrs)   # check against regex in case decimal number (e.g. '1.5' hours)
			if (dec_match):
				hrs = dec_match.group(1)   # convert into hours and minutes
				mins = str(int((float('0.'+dec_match.group(2)))*60))
		if (mins >= 60):
			print 'need to convert to hours!'
		print 'timeAllowed:', hrs, 'hours', mins, 'minutes'
	else:
		print 'sorry, couldn\'t find a suitable time from line:', line

	return

def parseCalc(line):   # function to interpret line containing calculator details, returning whether or not calculators are allowed

	allowed = ['allowed','permitted','can','may']
	negative = ['not','neither','nor']
	prohibited = ['prohibited','forbidden']

	calcsAllowed = None

	calcTokens = line.split()   # split line into individual word tokens

	for i in range(len(calcTokens)):   # traverse tokens
		if calcTokens[i] in prohibited:
			calcsAllowed = False
			break
		elif calcTokens[i] in allowed:
			if (calcTokens[i-1] not in negative) and ((i==len(calcTokens)-1) or (calcTokens[i+1] not in negative)):
				# don't check next token if at end of string
				calcsAllowed = True
				break   # no need to check further
			else:   
				# a negative word preceeds or suceeds an allowed word
				calcsAllowed = False
				break
	print 'calcsAllowed:', calcsAllowed

	return

if __name__=="__main__":   # entry point
   main()